docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,3025 @@ p4grammar {
+|   | $\vert$ | $\entry \pmatch \action$ | $\qquad \text{entries}$ |
+|   | $\vert$ | $\custom \bool \var \exp$ | $\qquad \text{custom property}$ |
+{ .booktable }
+
+**Methods:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\extMethod$ | $::=$ | $\constructor \var {\prmDef \baseTyp}$ | $\qquad \text{constructors}$ |
+|   | $\vert$ | $\abstractMethod \baseTyp \var \typVar {\prmDef \baseTyp}$ | $\qquad \text{abstract methods}$ |
+|   | $\vert$ | $\methodDef \baseTyp \var \typVar {\prmDef \baseTyp}$ | $\qquad \text{methods}$ |
+{ .booktable }
+
+**Declarations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\dcl$ | $::=$ | $\constDcl \baseTyp \var \exp$ | $\qquad \text{constants}$ |
+|   | $\vert$ | $\inst \baseTyp \args \var \stmts$ | $\qquad \text{instantiations (initialized)}$ |
+|   | $\vert$ | $\instNoBlock \baseTyp \args \var$ | $\qquad \text{instantiations (uninitialized)}$ |
+|   | $\vert$ | $\parserDcl \var \typVars {\overline { {\prmDef \baseTyp}}} {\overline {\VVal {\prmDef \baseTyp}}} \dcls \states$ | $\qquad \text{parsers}$ |
+|   | $\vert$ | $\controlDcl \var \typVars {\overline { {\prmDef \baseTyp}}} {\overline {\VVal {\prmDef \baseTyp}}} \dcls \stmts$ | $\qquad \text{controls}$ |
+|   | $\vert$ | $\funcDcl \baseTyp \var \typVars {\prmDefs \baseTyp} \stmts$ | $\qquad \text{functions}$ |
+|   | $\vert$ | $\externFuncDcl \baseTyp \var \typVars {\prmDefs \baseTyp}$ | $\qquad \text{extern functions}$ |
+|   | $\vert$ | $\varDclInit \baseTyp \var \exp$ | $\qquad \text{local variables (initialized)}$ |
+|   | $\vert$ | $\varDcl \baseTyp \var$ | $\qquad \text{local variables (uninitialized)}$ |
+|   | $\vert$ | $\valSetDcl \baseTyp \exp \var$ | $\qquad \text{parser value sets}$ |
+|   | $\vert$ | $\actionDcl \var {\prmDefs \baseTyp} \stmts$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\tableDcl \var \props$ | $\qquad \text{tables}$ |
+|   | $\vert$ | $\headerDcl \typVar \fieldTyps$ | $\qquad \text{header types}$ |
+|   | $\vert$ | $\headerUnionDcl \typVar \fieldTyps$ | $\qquad \text{header union types}$ |
+|   | $\vert$ | $\structDcl \typVar \fieldTyps$ | $\qquad \text{struct types}$ |
+|   | $\vert$ | $\errDcl \fields$ | $\qquad \text{error types}$ |
+|   | $\vert$ | $\matchkindDcl \fields$ | $\qquad \text{match kind types}$ |
+|   | $\vert$ | $\enumDcl \typVar \fields$ | $\qquad \text{enumerated types}$ |
+|   | $\vert$ | $\serEnumDcl \baseTyp \typVar \field \exp$ | $\qquad \text{serializable enumerated types}$ |
+|   | $\vert$ | $\externObjDcl \var \typVars \extMethods$ | $\qquad \text{extern objects}$ |
+|   | $\vert$ | $\typdefDcl \typVar \baseTyp$ | $\qquad \text{type definitions}$ |
+|   | $\vert$ | $\typdefDcl \typVar \dcl$ | $\qquad \text{type definitions (declarations)}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \baseTyp$ | $\qquad \text{generative type definitions}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \dcl$ | $\qquad \text{generative type definitions (declarations)}$ |
```

Comment:
IMO render newtype as [newtype X \rho] or for consistency with P4's concrete syntax [type X \rho]. The generative one shouldn't be an equation since X != \rho in the type system.

Reply:
shouldn't it be `type \rho t`? where `t` is a type identifier and not a type variable.

Reply:
Ah, yes. I always make this mistake.